src/Algebra/Graph/Relation/Internal.hs

```diff
@@ -164,15 +164,11 @@ overlay x y = Relation (domain x `union` domain y) (relation x `union` relation
 -- @
 connect :: Ord a => Relation a -> Relation a -> Relation a
 connect x y = Relation (domain x `union` domain y)
-    (relation x `union` relation y `union` (domain x `setProduct` domain y))
+    (relation x `union` relation y `union` (domain x `Set.cartesianProduct` domain y))
 
 instance NFData a => NFData (Relation a) where
     rnf (Relation d r) = rnf d `seq` rnf r `seq` ()
 
--- | Compute the Cartesian product of two sets. /Note: this function is for internal use only/.
-setProduct :: Set a -> Set b -> Set (a, b)
-setProduct x y = Set.fromDistinctAscList [ (a, b) | a <- Set.toAscList x, b <- Set.toAscList y ]
```

Comment:
It looks like we need to keep this function, but provide two different implementation: the current one for earlier GHC versions that do not ship with `containers-0.5.11`, and a new one based on `Set.cartesianProduct`. This will require some CPP, which is ugly, but seems unavoidable if we want to support earlier GHC versions.

Reply:
I will give this approach a try and get back to you!